### PR TITLE
Make registered more testable

### PR DESCRIPTION
**Pull request by lavalamp**
Pulling isolated changes out of my other PR. Should be a total no-op.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30663)

### CODE REVIEW

File: pkg/apimachinery/registered/registered.go [hunk 1]
```diff
@@ -32,8 +32,23 @@ import (
...
+	// should be registered in the scheme.
+	var err error
+	DefaultManager, err = NewManager(os.Getenv("KUBE_API_VERSIONS"))
+	if err != nil {
+		glog.Fatalf("Could not construct version manager: %v", err)
+	}
+}
+
+// Manager
+type Manager struct {
```
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: optional: Is RegisteredAPIsManager a better name?

File: pkg/apimachinery/registered/registered.go [hunk 2]
```diff
@@ -42,75 +57,101 @@ var (
...
 			}
-			envRequestedVersions = append(envRequestedVersions, gv)
+			m.envRequestedVersions = append(m.envRequestedVersions, gv)
 		}
 	}
+	return m, nil
 }
 
 // Resets everything to clean room for the start of a test
+// TODO: delete this
```
Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: Would be good to add details of why we want to delete this.

Perhaps say that testing should use its own instance of Manager and once we do that this should no longer be required.

**Comment by nikhiljindal** [comment]
Minor comments.
Feel free to self-apply the LGTM label after fixing.

File: pkg/apimachinery/registered/registered.go (hunk 1 above)
Comment (lavalamp) [inline pkg/apimachinery/registered/registered.go]: Changed to APIRegistrationManager

File: pkg/apimachinery/registered/registered.go (hunk 2 above)
Comment (lavalamp) [inline pkg/apimachinery/registered/registered.go]: Actually it took me like 10 minutes to just fix the tests. deleted.

Comment (nikhiljindal) [inline pkg/apimachinery/registered/registered.go]: awesome!

**Comment by lavalamp** [comment]
OK this push should also not crash on startup.

**Comment by lavalamp** [comment]
adding the label per reviewer :)

**Comment by MHBauer** [comment]
Make an instantiable object instead of a global static and then use that object. Seems pretty clear to me.

File: pkg/apimachinery/registered/registered.go [hunk 3]
```diff
@@ -42,75 +47,99 @@ var (
...
-	loadKubeAPIVersions()
+func NewOrDie(kubeAPIVersions string) *APIRegistrationManager {
+	m, err := NewAPIRegistrationManager(kubeAPIVersions)
+	if err != nil {
+		glog.Fatalf("Could not construct version manager: %v (KUBE_API_VERSIONS=%q)", err, kubeAPIVersions)
+	}
+	return m
 }
 
+// People are calling global functions. Let them continue to do that (for now).
```
Comment (MHBauer) [inline pkg/apimachinery/registered/registered.go]: Is there an issue for cleaning this up globally? Would be a good thing to track. [Maybe easy work for newbies?]

Comment (lavalamp) [inline pkg/apimachinery/registered/registered.go]: Not sure if there's an issue or not. I don't think it's easy work, unfortunately. :(

We *do* need a global registry for *available* apis, but this code should *not* be a global registry at all. Once we have the former, people will be able to make the latter at will. My https://github.com/kubernetes/kubernetes/pull/30565 is the next step.

**Pull request merged**